extensions-core/s3-extensions/src/main/java/io/druid/firehose/s3/StaticS3FirehoseFactory.java

```diff
@@ -22,54 +22,66 @@
 import com.fasterxml.jackson.annotation.JacksonInject;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
-import com.google.common.base.Charsets;
 import com.google.common.base.Preconditions;
-import com.google.common.base.Throwables;
-import com.google.common.collect.ImmutableList;
-import com.google.common.collect.Lists;
-
-import io.druid.data.input.Firehose;
-import io.druid.data.input.FirehoseFactory;
-import io.druid.data.input.impl.FileIteratingFirehose;
-import io.druid.data.input.impl.StringInputRowParser;
+import io.druid.data.input.impl.PrefetchableTextFilesFirehoseFactory;
 import io.druid.java.util.common.CompressionUtils;
+import io.druid.java.util.common.IAE;
 import io.druid.java.util.common.logger.Logger;
-
-import org.apache.commons.io.IOUtils;
-import org.apache.commons.io.LineIterator;
+import org.jets3t.service.S3ServiceException;
+import org.jets3t.service.ServiceException;
 import org.jets3t.service.impl.rest.httpclient.RestS3Service;
-import org.jets3t.service.model.S3Bucket;
 import org.jets3t.service.model.S3Object;
 
-import java.io.BufferedReader;
 import java.io.IOException;
 import java.io.InputStream;
-import java.io.InputStreamReader;
 import java.net.URI;
-import java.util.Iterator;
-import java.util.LinkedList;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
 import java.util.List;
+import java.util.stream.Collectors;
 
 /**
  * Builds firehoses that read from a predefined list of S3 objects and then dry up.
  */
-public class StaticS3FirehoseFactory implements FirehoseFactory<StringInputRowParser>
+public class StaticS3FirehoseFactory extends PrefetchableTextFilesFirehoseFactory<S3Object>
 {
   private static final Logger log = new Logger(StaticS3FirehoseFactory.class);
 
   private final RestS3Service s3Client;
   private final List<URI> uris;
+  private final List<URI> directories;
 
   @JsonCreator
   public StaticS3FirehoseFactory(
       @JacksonInject("s3Client") RestS3Service s3Client,
-      @JsonProperty("uris") List<URI> uris
+      @JsonProperty("uris") List<URI> uris,
+      @JsonProperty("directories") List<URI> directories,
```

Comment:
Could you call this "prefixes"? S3 doesn't really have directories (see http://docs.aws.amazon.com/AmazonS3/latest/UG/FolderOperations.html) so this term is misleading.